core/lcp/credential.py

```diff
@@ -13,6 +16,36 @@ class LCPCredentialType(Enum):
     LCP_HASHED_PASSPHRASE = "Hashed LCP Passphrase passed to the LCP License Server"
 
 
+class LCPHashedPassphrase(object):
+    """A hashed passphrase."""
+
+    hashed: str
```

Comment:
This is just an attribute declared with a type and without an initial value. I spoke with JG privately a few weeks ago and this seemed to be the way we'd like to do things: All attributes declared at the top of the class with type annotations. There are a lot of places in the codebase where attributes are conjured up out of thin air in the middle of a class, and nobody has any idea what the type should be.

Reply:
For non-`dataclass`es or similar (e.g., pydantic BaseModel), a declaration here WITH a value would be a class variable. I don't know exactly what is happening here, but it may not be exactly what it looks like.

I totally agree that new instance variables should not be created outside of the instance initializer, in most cases. The usual way to declare the instance variables would be in the `__init__` function with `self.variable_name = ...` or `self.variable_name: TYPE = ...`.

I truly didn't know that you could even do this without something like this:
```
from dataclasses import dataclass
...
@dataclass
class ClassName:
    ...
```

Reply:
It's something that's unique to MyPy:

https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes

The declaration serves as a way to state the type regardless of whether the actual attribute ends up as an instance or class attribute.

Reply:
Thanks! That was the clue I needed. Reviewing the docs, declaring here will prevent MyPy from distinguishing between class and instance variables, so it still may be preferable to declare in the `__init__` method.

Reply:
The intention seems to be that you add those annotations unless you know that something could be a class attribute, in which case you'd use `ClassVar[T]` to actually enforce it.

Reply:
This is what I understand from the docs...

```
class SomeClass:
    var: T  # class or instance variable
```

```
class SomeClass:
    var: ClassVar[T]  # class variable only
```

```
class SomeClass:
    def __init__(self, var: T) -> None:
        self.var = var  # instance variable only (inferred from signature)
```

```
class SomeClass:
    def __init__(self) -> None:  # instance variable only
        self.var: T = ...   # instance variable only
```

Reply:
There's something I don't like about the constructor approach that a small example like this won't show: You're required to follow the execution of lines of code to determine what type an attribute will end up being. That's backwards from a theoretical perspective. We do have quite a few constructors in the codebase that are of a decent enough length that this is a pain. Much less pain involved to just read a list of attributes declared up front.

This is not a practical problem where there's a 1:1 mapping between attributes and constructor parameters, but that's not even the common case in this codebase, sadly.